Synchronization

Sairajkodilkar opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): x86_64, arm64
 - Platform(s): N/A
 - Application(s): N/A


<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->


<!--
Please provide a detailed description of the changes made in this new PR.
-->
### Description

The pull request introduces irq save API for the spinlock as well as SMP safe mutex implementation.
Previously for UP arch, the Mutex disabled the interrupt, this implementation also synchronized the wait queue. 
Along with this the previous implementation also used the two conditions to indicate the lock that lock is free, which was
`owner == current || lock_count == 0
`
but if we observe closely we can separate these conditions since if the owner is the current thread then we can increment the lock count, otherwise, we can wait for the current to be null and can acquire the lock. This implementation does not disable interrupts but we can definitely introduce an API that does. 
One requirement of this commit is that we need to separately implement the SMP safe wait queue. currently, these wait queue is only used in uklock lib but SMP safe wait queues can be used anywhere in the kernel.

marcrittinghaus reviewed (commented):
Hi @Sairajkodilkar,

Thanks for supplying the PR 😃 Just a few comments:
- Since it is clear from the `lock`/`unlock` what operation this entails for the IRQ flags, I would just name the macros `uk_spin_lock_irq`, `uk_spin_unlock_irq` etc.
- As we discussed on Discord, not all spinlocks require that we save/restore IRQ flags. To accommodate for that matter I suggest to keep the flags outside of the spinlock as a stack variable just like before. So the new macros `uk_spin_lock_irq` should have a second parameter for the variable which should receive the flags. That will save memory for all other spinlocks.
- Is there a particular reason that you use `sub_fetch` with `1` instead of just `dec` etc.? Have you checked the ASM output? I am asking because a `lock dec` would be shorter to encode than a `lock sub`.

Sairajkodilkar commented:
Hey, @marcrittinghaus thanks for the review

- I thought it would be cleaner to store the irqf with the spinlock, but since we have memory constraints it would be better to use the stack variable.
- The reason behind using the sub_fetch is that it returns the value after decrementing the count, whereas dec returns the value and then performs the decrement operation.
- I did not understand why dec would result in a shorter encoding because the dec is a macro that expands to atomic_fetch_sub()

marcrittinghaus commented:
> * I did not understand why dec would result in a shorter encoding because the dec is a macro that expands to atomic_fetch_sub()

Ah, yeah, you are right. I was thinking about the x86 assembler and did not lookup the macro definition 👍🏻

marcrittinghaus commented:
Had a quick look at the code and it looks good so far 😃 Great work @Sairajkodilkar! Of course, we need to have full reviews in the end to verify correctness and details...

marcrittinghaus commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,304 @@
+
+/* FIXME: lost wakeup problem
+ *			one solution: use deadline base waking up
+ */
```
Hi @Sairajkodilkar,

is there still a bug here? What is the problem?

Sairajkodilkar replied:
I resolved the bug in the commit ae96713a5c8cf2535fe754d87ac1937f753f579e. But I have not tested the code.

marcrittinghaus commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,304 @@
+		uk_waitq_wait_event(&rwl->shared, _rw_can_read(rwl->rwlock));
+	}
+
+	return;
```
Remove the trailing returns as we do not have them anywhere else.

marcrittinghaus commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,304 @@
+	uintptr_t rwlock;
+#ifdef CONFIG_RWLOCK_WRITE_RECURSE
+	unsigned int write_recurse;
+#endif
```
```suggestion
#endif /* CONFIG_RWLOCK_WRITE_RECURSE */
```

Always add a comment to the endif with the condition

marcrittinghaus commented on lib/uklock/include/uk/arch/arm64/ticketlock.h:
```diff
@@ -54,6 +54,7 @@ static inline void ukarch_ticket_init(struct __ticketlock *lock)
 {
 	lock->next = 0;
 	lock->current = 0;
+	lock->irqf = 0;
```
Is this still needed?

Sairajkodilkar replied:
No, We don't need it, I created a separate interface for irq enable disable.

marcrittinghaus commented on lib/uklock/Makefile.uk:
```diff
@@ -5,3 +5,4 @@ CXXINCLUDES-$(CONFIG_LIBUKLOCK)	+= -I$(LIBUKLOCK_BASE)/include
 
 LIBUKLOCK_SRCS-$(CONFIG_LIBUKLOCK_SEMAPHORE) += $(LIBUKLOCK_BASE)/semaphore.c
 LIBUKLOCK_SRCS-$(CONFIG_LIBUKLOCK_MUTEX)     += $(LIBUKLOCK_BASE)/mutex.c
+LIBUKLOCK_SRCS-$(CONFIG_LIBUKLOCK_RWLOCK)     += $(LIBUKLOCK_BASE)/rwlock.c
```
There seems to be a space too many.

marcrittinghaus commented on include/uk/arch/atomic.h:
```diff
@@ -56,6 +56,14 @@ extern "C" {
 	__atomic_fetch_add(src, value, __ATOMIC_SEQ_CST)
 
+/**
+ * Perform a atomic add/sub and fetch operation
```
```suggestion
 * Perform an atomic add/sub and fetch operation
```

andraprs replied:
With regard to the commit message:

"Unikraft provides macros for limited number ..." => "Unikraft provides macros for a limited number ...".

For the functions comment, in general, can end the sentence with a dot, if that is the coding style from the rest of the file.  For example, including the above suggested change from Marc:

"Perform a atomic add/sub and fetch operation" => "Perform an atomic add/sub and fetch operation."

marcrittinghaus commented on arch/x86/x86_64/include/uk/asm/spinlock.h:
```diff
@@ -39,6 +39,7 @@ struct __spinlock {
 };
 
 /* Initialize a spinlock to unlocked state */
+/* TODO: discuss the use of of this initializer */
```
Let us discuss. What was your question here?

Sairajkodilkar replied:
Hi @marcrittinghaus, It was a simple discussion in the early stage of development. I just remembered that I need to remove this and other TODOs.

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -98,53 +100,71 @@ void uk_mutex_get_metrics(struct uk_mutex_metrics *dst);
+		 * Here the owner act as the lock
+		 * Here We dont need to disable interrupts because the compare exchange
+		 * is atomic even on the single processor
+		 * TODO: discuss the possibility that interrupt handler may require the
+		 * same lock, resulting in deadlock.
```
My gut feeling here is that disabling (the) IRQ should be done by the user if needed and not generally in every case.

But if you have something like:
```C
flags = ukplat_lcpu_save_irqf();
uk_mutex_lock(&mutex);
/* do something */
uk_mutex_unlock(&mutex);
ukplat_lcpu_restore_irqf(flags);
```
and you don't get the mutex and would block, it is really bad that IRQs are turned off as long as you block (which might actually also deadlock the system).

Nevertheless, acquiring a mutex in an IRQ handler is not a great idea anyways and the proper way would be to use a spinlock with `uk_spin_lock_irqf()` instead.

Sairajkodilkar replied:
Hi @marcrittinghaus, you are right that it is the programmer's job not to sleep while holding a lock or not hold a lock which is needed by an interrupt handler.

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -98,53 +100,71 @@ void uk_mutex_get_metrics(struct uk_mutex_metrics *dst);
-			m->lock_count == 0 || m->owner == current);
-		irqf = ukplat_lcpu_save_irqf();
-		if (m->lock_count == 0 || m->owner == current)
+		/* TODO: make wait queue SMP safe */
```
Looking at the changes to the waitqueue below, I guess this is done by now?

Sairajkodilkar replied:
yes I added the a lock for a queue, but in future we need a SMP safe threads and scheduler

marcrittinghaus commented on lib/uklock/Config.uk:
```diff
@@ -38,4 +46,18 @@ if LIBUKLOCK
 			objects, as well as number of successful/failed locking attempts
 			since startup.
+
+	config LIBUKLOCK_MUTEX_RECURSE
+		bool "Recursive mutex lock"
+		default n
+		depends on LIBUKLOCK_MUTEX
+		help
+			Enable recursive locking for mutex
+
+	config LIBUKLOCK_RWLOCK_WRITE_RECURSE
+		bool "Recursive writer lock"
+		default n
+		depends on LIBUKLOCK_RWLOCK
+		help
+			Enable recursive writer lock
```
I am wondering if it makes sense to globally enable/disable this. Because if you want to use it in a single place this will affect all locks in the system, which might also hide bugs or trigger unexpected or unintended behavior.

Would it be possible to decide on a per-lock basis, when initializing the lock?

Sairajkodilkar replied:
Yes, we can keep a write recurse flag along with rwlock. This flag can be configured at the time of initialization.

marcrittinghaus commented on lib/uksched/include/uk/thread.h:
```diff
@@ -106,6 +106,15 @@ struct uk_thread *uk_thread_current(void)
 	return *current;
 }
 
+static inline
+unsigned long uk_get_stack_bottom(void)
+{
+	unsigned long sp = ukarch_read_sp();
+
+	return sp & STACK_MASK_TOP;
+}
+
+
```
Regarding changes in `uksched` we should wait for the sched/musl changes being upstream and then adapt as necessary.

Sairajkodilkar replied:
Yes I agree, do you have any better alternative for the owner, like a property of thread whose value is unique per thread ?

marcrittinghaus commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,304 @@
+
+void uk_rwlock_init(struct uk_rwlock *rwl);
+
+static inline
+void uk_rwlock_rlock(struct uk_rwlock *rwl)
```
I am wondering if it really makes sense to still have these functions as `static inline`, as they are much longer than for mutex/semaphore.

Sairajkodilkar replied:
I did this because mutex and semaphore were defined in header files.
I am planning to split the rwlock in the header and `.c` file.

marcrittinghaus reviewed (commented):
Hi @Sairajkodilkar,

thanks for putting together this PR! I have some first feedback and also some questions 😃

andraprs commented on arch/arm/arm64/include/uk/asm/spinlock.h:
```diff
@@ -39,6 +39,7 @@
 
```
With regard to the commit messages, in general, can you please double-check them and, where it is needed, can you please adjust the text e.g. remove duplicate wording, focus more on the why & what for the proposed changes, can replace "This commit <adds / provides / other action> ..." with "Add / Provide / Other action ...", fix typos. If not already aware of this reference - https://cbea.ms/git-commit/, could check this out.

For example: "Add a spinlock API to save and restore the irq flags".

Also, can you please include in the PR overview description the details of the testing that you've done so far (e.g. scenarios, commands used).

Thank you.

andraprs commented on arch/arm/arm64/include/uk/asm/spinlock.h:
```diff
@@ -39,6 +39,7 @@
 
 struct __align(8) __spinlock {
 	volatile int lock;
+	unsigned int irq;
```
Shouldn't this be "irqf", similar to the other changes in this commit?

Same for the change below.

andraprs commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -98,52 +98,63 @@ void uk_mutex_get_metrics(struct uk_mutex_metrics *dst);
+		 * lock count
+		 *
+		 * Here the owner act as the lock
+		 * Here We dont need to disable interrupts because the compare exchange
```
Can remove "Here" from the beginning of this sentence.

andraprs commented on arch/arm/arm64/include/uk/asm/spinlock.h:
```diff
@@ -43,7 +43,7 @@ struct __align(8) __spinlock {
 };
```
As per the initial comment from my side, could modify, for example, this commit message to "Update the UKARCH_SPINLOCK_INITIALIZER macro to set the irq flags to 0.".

andraprs commented on lib/uklock/include/uk/spinlock.h:
```diff
@@ -75,26 +75,26 @@ extern "C" {
 }
 
 #define uk_spin_trylock_irqf(lock) { \
 	lock->irqf = ukplat_lcpu_save_irqf(); \
```
Here should be "(lock)->irqf", as per the other changes in this commit, for these macros.

andraprs commented on lib/uksched/include/uk/wait.h:
```diff
@@ -93,6 +98,10 @@ do { \
 	ukplat_lcpu_restore_irqf(flags); \
 } while (0)
 
+/* TODO investigate uk_sched_thread_blocked
```
Please let us know if there are any updates about this.

andraprs commented on arch/x86/x86_64/include/uk/asm/spinlock.h:
```diff
@@ -36,7 +36,6 @@
 
 struct __spinlock {
 	volatile int lock;
-	unsigned int irqf;
```
Can you please squash the following commit "arch/\*, lib/uklock, include/uk/arch/spinlock.h: Change the spinlock IRQ API" with "arch/\*, lib/uklock: Add save irqf API for the spinlock ", "arch/*: Remove irqf initialization from uk_spin_init ", and keep the latest version of the API.

andraprs commented on lib/uklock/include/uk/semaphore.h:
```diff
@@ -78,22 +79,22 @@ static inline int uk_semaphore_down_try(struct uk_semaphore *s)
 #ifdef UK_SEMAPHORE_DEBUG
 		uk_pr_debug("Decreased semaphore %p to %ld\n",
-			    s, s->count);
+				s, s->count);
```
Unrelated change. Same for the ones below, if they are not only shown by the codebase preview on the web page and they are actual changes.

andraprs commented on lib/uklock/include/uk/semaphore.h:
```diff
@@ -78,22 +79,22 @@ static inline int uk_semaphore_down_try(struct uk_semaphore *s)
 /* Returns __NSEC_MAX on timeout, expired time when down was successful */
 static inline __nsec uk_semaphore_down_to(struct uk_semaphore *s,
-					  __nsec timeout)
+		__nsec timeout)
```
Unrelated change.

andraprs commented on lib/uklock/include/uk/semaphore.h:
```diff
@@ -105,23 +106,23 @@ static inline __nsec uk_semaphore_down_to(struct uk_semaphore *s,
+		irqf = uk_spin_lock_irqf(&(s->sl));
 		if (s->count > 0 || (deadline &&
-				     ukplat_monotonic_clock() >= deadline))
+					ukplat_monotonic_clock() >= deadline))
```
Unrelated change.

andraprs commented on lib/uklock/include/uk/semaphore.h:
```diff
@@ -105,23 +106,23 @@ static inline __nsec uk_semaphore_down_to(struct uk_semaphore *s,
 #ifdef UK_SEMAPHORE_DEBUG
 		uk_pr_debug("Decreased semaphore %p to %ld\n",
-			    s, s->count);
+				s, s->count);
```
Unrelated change.

andraprs commented on lib/uklock/include/uk/semaphore.h:
```diff
@@ -134,14 +135,14 @@ static inline void uk_semaphore_up(struct uk_semaphore *s)
 #ifdef UK_SEMAPHORE_DEBUG
 	uk_pr_debug("Increased semaphore %p to %ld\n",
-		    s, s->count);
+			s, s->count);
```
Unrelated change.

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,29 @@
+#ifndef __UK_RWLOCK_H__
```
Should include the license / copyright header(s) in all new files. One example of the header can be found in: https://github.com/unikraft/unikraft/blob/staging/lib/uklock/include/uk/spinlock.h.

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -60,7 +60,7 @@ static inline void uk_rwlock_rlock(struct uk_rwlock *rwl)
 }
 
-static inline void uk_rwlock_runlock(struct uk_rwlock *rwl)
+static inline void uk_rwlock_wlock(struct uk_rwlock *rwl)
```
Can squash commit "lib/uklock: Add lock function for reader and writer" with "lib/uklock: Correct the function name", to have the corresponding function names in one commit.

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -24,10 +24,6 @@
 #define OWNER(x)					(x & ~FLAG_MASK)
 
 struct uk_rwlock {
-	/* TODO: discuss that should we store struc thread ** since it would give us
```
Please include a commit message description for the following commit: "lib/uklock: Add the priority for writer locks".

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -37,8 +38,10 @@ void uk_rwlock_init(struct uk_rwlock *rwl);
 static inline 
 bool _rw_can_read(unsigned int rwlock)
 {
+	/* Give priority to the writers */
```
Should explain in the commit message which is the bug that is fixed in the following commit: "lib/uklock: Resolve the bug in reader writer unlock".

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -9,42 +9,48 @@
 
-#define READERS_SHIFT				(5)
-#define FLAG_MASK					(~(1 << READERS_SHIFT))
+#define FLAG_BITS		(5)
```
Should explain the fixed bugs and cleanups in this commit " lib/uklock: Complete the implementation of lock and unlock for reader-writer".

Or, better, put together all these commits for read/write locks, the initial implementation and the fixes / cleanups. Then can split in different commits the functionality for read / write lock / unlock, that includes the latest version of the codebase, with the fixes.

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -1,11 +1,15 @@
- *	Is reader or Is writer
- */
+#include <uk/config.h>
+
```
Should include the changes from the commit "lib/uklock: Add the upgrade and downgrade functions for reader writer
lock", that are not related to the upgrade / downgrade functionality, in the corresponding commits for lock / unlock.

andraprs commented on include/uk/arch/atomic.h:
```diff
@@ -72,6 +72,13 @@ extern "C" {
 	ukarch_fetch_add(src, 1)
 #define ukarch_dec(src) \
 	__atomic_fetch_sub(src, 1, __ATOMIC_SEQ_CST)
+
```
The following commit "lib/uklock, include/uk/arch/atomic.h: Add the atomic OR and remove bug in downgrade" should be split and keep the "OR" functionality in a separate commit. Then include the bug fixes in the corresponding commit for the rw lock downgrade functionality.

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -10,12 +10,14 @@
 #define FLAG_BITS		(5)
 #define READERS_SHIFT		(FLAG_BITS)
-#define FLAG_MASK		(~(1 << READERS_SHIFT))
+#define FLAG_MASK		((1 << READERS_SHIFT) - 1)
```
Please include the fixes from the commit "lib/uklock: Remove bugs in reader-writer lock" in the corresponding commits for each functionality e.g. rw lock / unlock / upgrade / downgrade.

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -49,25 +48,24 @@ struct __align(8) uk_rwlock {
 void uk_rwlock_init(struct uk_rwlock *rwl);
 
-static inline 
+static inline
```
Please include the fixes for the issues identified by checkpatch in the initial commits that introduced each functionality.

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -233,7 +242,7 @@ void uk_rwlock_upgrade(struct uk_rwlock *rwl)
 		/* If we cannot upgrade wait till readers count is 0 */
 		ukarch_or(&rwl->rwlock, UK_RWLOCK_WRITE_WAITERS);
-		uk_waitq_wait_event(&rwl->exclusive, _rw_can_write(rwl->rwlock));
+		uk_waitq_wait_event(&rwl->exclusive, _rw_can_upgrade(rwl->rwlock));
```
Similarly as before, should include the fixes in the initial commits for the upgrade / downgrade path.

andraprs commented on lib/uklock/Config.uk:
```diff
@@ -46,4 +46,18 @@ if LIBUKLOCK
 			objects, as well as number of successful/failed locking attempts
 			since startup.
+
+	config LIBUKLOCK_MUTEX_RECURSE
```
The following commit "lib/uklock: Add recursive lock configuration" could be split in two commits, one for the mutex, one for the rw lock.

andraprs commented on lib/ukrust/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+[package]
```
This commit - "lib/ukrust: Update ukrust to work alongside Cargo" - and the previous one "lib/nolibc: Add bcmp" don't seem related to the PR topic, possibly they are part of a rebase.

andraprs reviewed (commented):
Added a set of high level comments, as part of the initial review.

Thank you.

Sairajkodilkar committed:
include/arch: Add macros for atomic add, sub, or

This commit adds macros to atomically add, subtract, and
perform a bitwise or based on the builtins provided by GCC.

Signed-off-by: Sairaj Kodilkar <skodilkar7@gmail.com>

Sairajkodilkar committed:
lib/uklock: Provide irqf wrappers for uk_spinlock

Holding a spinlock which is required by an interrupt handler, causes a
deadlock. To avoid this, the programmer must disable the interrupt
before acquiring such spinlocks. This involves saving the interrupt
flags before acquiring the spinlock and restoring them after
releasing the spinlock.

This commit provides versions of the ukplat_spin_* macros for the
uk_spinlock.

Co-authored-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>
Signed-off-by: Sairaj Kodilkar <skodilkar7@gmail.com>
Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

Sairajkodilkar committed:
lib/uklock: Make mutex SMP safe

We can safely acquire a lock on uniprocessor systems by disabling the
interrupts, but this does not work in an SMP environment. Hence, to
make the mutex SMP safe, this commit uses atomic compare exchange.
The commit also makes recursive locking optional (disabled by default).
While this is a breaking change, recursive locks are discouraged and
should only be used with caution. The current Unikraft core does not
use recursive locks anywhere.

Co-authored-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>
Signed-off-by: Sairaj Kodilkar <skodilkar7@gmail.com>
Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

Sairajkodilkar committed:
lib/uklock: Make semaphore SMP safe

On uniprocessor systems, it is sufficient to disable interrupts.
In SMP environments, however, it is not because a thread executing on a
different core can give rise to a race condition. Hence, to make the
semaphore SMP safe, we use a spinlock.

Signed-off-by: Sairaj Kodilkar <skodilkar7@gmail.com>

Sairajkodilkar committed:
lib/uksched: Make wait queues SMP safe

Wait queues use a linked list to store the threads waiting on a
particular condition. But operations such as add to and remove
from the list are not SMP safe. Hence, this commit associates
a spin lock with each wait queue.

Co-authored-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>
Signed-off-by: Sairaj Kodilkar <skodilkar7@gmail.com>
Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

Sairajkodilkar commented:
Hi @andraprs and @marcrittinghaus, I have restructured the commit history. Please have a second look at it
Also, @andraprs, Can you please give me some time to write basic tests? so that it will be easier for you as well as @marcrittinghaus to test

andraprs commented:
> Hi @andraprs and @marcrittinghaus, I have restructured the commit history. Please have a second look at it Also, @andraprs, Can you please give me some time to write basic tests? so that it will be easier for you as well as @marcrittinghaus to test

Thank you for the updates, Sairaj. :) I'll go through the patch series during this week and next week.

Sure, that's fine, if you would like to also add a set of basic tests. One option can be using "uktest", as Marc mentioned in the Unikraft SMP community meeting. Part of my next steps is to get into more details about the testing setup (e.g. using "uktest"), I can also help with this.

andraprs commented on include/uk/arch/atomic.h:
```diff
@@ -64,6 +72,13 @@ extern "C" {
 	__atomic_fetch_sub(src, 1, __ATOMIC_SEQ_CST)
+
+/**
+ * Perform the atomic OR operation and returns the old value
```
"Perform the atomic OR operation and returns the old value" => "Perform an atomic OR operation and return the old value."

andraprs commented on lib/uklock/include/uk/spinlock.h:
```diff
@@ -74,6 +74,25 @@ extern "C" {
 
 #endif	/* CONFIG_LIBUKLOCK_TICKETLOCK */
 
+#define uk_spin_lock_irqf(lock) ({					\
```
With regard to the commit message:

"release the spinlock and set the interrupt flag to the value provided" => "Release the spinlock and set the interrupt flag to the provided value."

andraprs commented on lib/uksched/include/uk/thread.h:
```diff
@@ -97,6 +97,14 @@ struct uk_thread *uk_thread_current(void)
 	return __uk_sched_thread_current;
 }
 
+static inline
```
With regard to the commit message:

"Also stack is aligned to the 4K boundary, this allows us to use the
bottom 12 bit of the stack pointer to store lock specific info such as
flags" => "Also stack is aligned to the 4K boundary. This allows us to use the
bottom 12 bit of the stack pointer to store lock specific info such as
flags."

andraprs commented on lib/uksched/include/uk/thread.h:
```diff
@@ -137,6 +145,8 @@ struct uk_thread *uk_thread_current(void)
  *       termination callbacks as well.
  */
 
+
```
There seems to be two empty lines here, unrelated to the current commit.

andraprs commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -98,53 +100,68 @@ void uk_mutex_get_metrics(struct uk_mutex_metrics *dst);
 
 	current = uk_thread_current();
 
+	/* if the owner is current thread then it definitely can't unlock during
```
The comment could be split in two. For example: "If the owner is the current thread, then it definitely can't unlock during uk_mutex_lock. Then we can simply increment the lock count.".

andraprs commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -98,53 +100,68 @@ void uk_mutex_get_metrics(struct uk_mutex_metrics *dst);
-		if (m->lock_count == 0 || m->owner == current)
+		uk_waitq_wait_event(&m->wait, m->owner == NULL);
+
+		/* If there is no owner for the lock we can grab it and increment the
```
It depends on the coding style so far and what corresponds to each file. Could add a dot after each sentence, as a delimiter. For example: "If there is no owner for the lock we can grab it and increment the lock count. The owner acts as the lock. We don't need to disable interrupts because the compare exchange is atomic even on the single processor."

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,306 @@
+		return;
+#endif
+
+	/* All recusive locks have been released, time to unlock*/
```
"recusive" => "recursive"

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,306 @@
+	stackbottom = uk_get_stack_bottom();
+	v = rwl->rwlock;
+
+	UK_ASSERT(rwl);
```
The assert needs to be moved after the local variable declarations, before accessing "rwl".

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,306 @@
+#ifndef __UK_RWLOCK_H__
```
Should include the license / copyright header(s) in all new files, such as "rwlock,h" and "rwlock.c". One example of the header can be found in: https://github.com/unikraft/unikraft/blob/staging/lib/uklock/include/uk/spinlock.h.

andraprs commented on lib/uklock/include/uk/spinlock.h:
```diff
@@ -74,6 +74,25 @@ extern "C" {
 
 #endif	/* CONFIG_LIBUKLOCK_TICKETLOCK */
```
What about the changes for the "ticketlock" from the previous revision?

Sairajkodilkar replied:
Hi, Previous changes stored the interrupt flag with the lock. The new implementation provides a different interface for IRQF store and restores. Hence I restored the changes in ticketlock

andraprs commented on lib/uklock/Config.uk:
```diff
@@ -30,6 +31,13 @@ if LIBUKLOCK
+		select LIBUKSCHED
+		default y
+		help
+			Enable Reader-Writer based synchornization
```
"synchornization" => "synchronization"

andraprs commented on include/uk/arch/atomic.h:
```diff
@@ -56,6 +56,14 @@ extern "C" {
 #define ukarch_fetch_add(src, value) \
 	__atomic_fetch_add(src, value, __ATOMIC_SEQ_CST)
 
```
For the overall changes from the PR, please also check the Codacy results [1] and the modifications / tags that are needed based on the "checkpatch" warnings.

[1] https://github.com/unikraft/unikraft/pull/476/checks

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -14,7 +14,7 @@
 
 /* TODO: in future give spin as well as wait options to the user */
 
```
With regard to the commit message: 

"Decrease number of reader-writer flags from five" => "Decrease number of reader-writer flags bits from five"

"in the lower five bits
owner field, This was possible because stack was align to 4K and we had
tweleve free bits in stack pointer. " => "in the lower five bits
owner field. This was possible because the stack was aligned to 4K and we had
twelve free bits in the stack pointer. ".

Can you also please reference in the description the commit that includes the alignment change (e.g. add commit title and commit id).

Sairajkodilkar replied:
Hi, I think this comment belongs to the other commit.

andraprs commented on lib/uklock/exportsyms.uk:
```diff
@@ -4,3 +4,9 @@ uk_mutex_get_metrics
 _uk_mutex_metrics
 _uk_mutex_metrics_lock
 uk_rwlock_init
+uk_rwlock_rlock
```
With regard to the commit message: "Hence this commits move these functions from header file to
source file" => "Hence this commit moves these functions from the header file to the
source file".

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -14,7 +14,7 @@
 /* TODO: in future give spin as well as wait options to the user */
 
-#define FLAG_BITS		(5)
+#define FLAG_BITS		(4)
```
Just double-checking, is there any other update needed for the current implementation given this bits number decrease? (e.g. update codebase that relies on having 5 bits available)

Sairajkodilkar replied:
No, current implementation uses only 3 bits. The remaining bits are reserved for future use.

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -40,29 +40,24 @@
 struct __align(8) uk_rwlock {
 	uintptr_t rwlock;
-#ifdef CONFIG_RWLOCK_WRITE_RECURSE
+#define UK_RW_CONFIG_SPIN			0x01
```
For the flags name, it should be RWLOCK instead of RW. Just to target the type of lock, as it was initially done for the Kconfig options.

And the flags defines should be moved outside of the data structure definition, in a separate codebase block (e.g. before "struct __align(8) uk_rwlock").

andraprs commented on lib/uklock/Config.uk:
```diff
@@ -53,11 +53,4 @@ if LIBUKLOCK
 		depends on LIBUKLOCK_MUTEX
 		help
 			Enable recursive locking for mutex
-
```
The commit message could be updated to "Remove compile time flags since the configuration flags are set during runtime." .

andraprs commented on lib/uklock/exportsyms.uk:
```diff
@@ -1,5 +1,5 @@
 uk_semaphore_init
-uk_mutex_init
+__uk_mutex_init
```
In the commit message, there should be "UK_MUTEX_CONFIG_WRITE_RECURSE" instead of "UK_RW_CONFIG_WRITE_RECURSE".

andraprs commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -52,18 +52,21 @@
-#ifdef CONFIG_LIBUKLOCK_MUTEX_RECURSE
 	int lock_count;
-#endif
+#define UK_MUTEX_CONFIG_WRITE_RECURSE 0x01
```
Similarly in this case, the flag define should be moved outside of the data structure definition, in a separate codebase block.

andraprs commented on lib/uklock/Config.uk:
```diff
@@ -46,11 +46,4 @@ if LIBUKLOCK
 			Metrics related to mutex objects: current amount of (un)locked
 			objects, as well as number of successful/failed locking attempts
 			since startup.
-
```
As for the previous occurrence, the commit message could be "Remove compile time flags since the configuration flags are set during runtime." .

andraprs commented on lib/uklock/include/uk/lock-common.h:
```diff
@@ -0,0 +1,31 @@
+#include <uk/plat/lcpu.h>
```
This new file also should have the license / copyright header(s).

And a couple of notes for the commit message:

* "the lock type and function as argument. The macro then define" => "the lock type and function as arguments. The macro then defines".

* "Defines a new function which saves irq flag before acquiring lock" => "Defines a new function which saves the irq flag before acquiring lock." (just a "." at the end, the same as for the other bullet points sentences).

* "before trying to aquire the flags" => ""before trying to acquire the flags"

* "if the lock is not aquired." => "if the lock is not acquired."

andraprs commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -52,6 +52,7 @@
 extern "C" {
 #endif
 
+#include <uk/lock-common.h>
```
Should mention in the commit message the lock types e.g. "This commit defines the irq safe interface for the mutex, rwlock and spinlock.". As, for example, the semaphore is not part of this commit.

"trying to aquire a lock inside a interrupt" => "trying to acquire a lock inside an interrupt"

andraprs commented on lib/uklock/include/uk/spinlock.h:
```diff
@@ -37,6 +37,10 @@
 #endif
 
+#include <uk/lock-common.h>
+#define _IRQFNAME(funcname) funcname##_irqf
```
This should be removed, given it is already defined in the "lock-common" header.

andraprs commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -27,7 +27,6 @@
 #define UK_RW_READERS(x)			((x) >> READERS_SHIFT)
 
 #define UK_RW_UNLOCK				(UK_RW_READERS_LOCK(0))
-#define OWNER(x)					((x) & ~FLAG_MASK)
```
A couple of notes for the commit message:

* "the owner and flags for the reader writer lock were store" => "the owner and flags for the reader writer lock were stored"
* Could add in the commit description a reference to the commit that updated the stack alignment e.g. commit title and id.
* "separates the owner and flag field" => "separates the owner and flag fields"
* "new implementation uses pointer to the `struct uk_thread` as the owner of the lock" => "the new implementation uses a pointer to the `struct uk_thread` as the owner of the lock."

andraprs commented on lib/uklock/rwlock.c:
```diff
@@ -1,3 +1,4 @@
+#include <uk/thread.h>
```
Should be included in the "rwlock" header file, as "struct uk_thread" is used there as a struct field type.

andraprs reviewed (commented):
Thank you for the updated version of the PR. It has been smoother to go through the patch series.

I added a set of questions and suggestions. In general, the locking primitives proposals look good. Thank you.

Sairajkodilkar committed:
lib/uksched: Add unconditional wait and wakeup_one

This commit adds an unconditional wait, which is required
for reader-writer locks. The commit also adds an interface to
wakeup only a single thread sleeping on a wait queue.

Co-authored-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>
Signed-off-by: Sairaj Kodilkar <skodilkar7@gmail.com>
Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

Sairajkodilkar committed:
lib/uklock: Add SMP-safe reader-writer lock

This commit introduces a reader-writer lock. To avoid
starvation of readers and writer, the lock maintains the current
count of readers and writers waiting for the lock. New readers give
priority to waiting writers and writers give priority to readers on
unlock.

Co-authored-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>
Signed-off-by: Sairaj Kodilkar <skodilkar7@gmail.com>
Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

andraprs reviewed (approved):
Thank you for the updates from the latest revision and for developing and sharing a testing setup [1]. I checked out the patch series using the provided test app.

[1] https://github.com/Sairajkodilkar/unikraft/commits/sched-sync

marcrittinghaus commented on lib/uklock/include/uk/lock-common.h:
```diff
@@ -0,0 +1,63 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
```
The NEC part of the header does not make sense 😏 You can just use this one from the new coding conventions to simplify and also save some trees for people printing the code 😄🌲 

```
/* SPDX-License-Identifier: BSD-3-Clause */
/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
 * Licensed under the BSD-3-Clause License (the "License").
 * You may not use this file except in compliance with the License.
 */
```

marcrittinghaus commented on lib/uklock/include/uk/lock-common.h:
```diff
@@ -0,0 +1,63 @@
+	{															\
+		funcname(lock);											\
+		ukplat_lcpu_restore_irqf(irqf);							\
+	}
```
Please ensure that you use 8 spaces for each tab. The `\` should be in column 73 on all lines.

marcrittinghaus commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,87 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
```
See header comment.

marcrittinghaus commented on lib/uklock/include/uk/semaphore.h:
```diff
@@ -1,5 +1,7 @@
 /* SPDX-License-Identifier: BSD-2-Clause */
 /*
+ * Authors: Sairaj Kodilkar <skodilkar7@gmail.com>
+ *
```
Please refrain from adding yourself as author for the existing files, if changes are marginal (e.g., if you did not completely change how things are done or added major things). Generally, we would like to get away from the authors field and use the shorter more generic header as stated in the earlier comment. The authors field is quickly outdated and does not provide information on what changes you have authored (e.g., in this case, it suggests you have introduced the whole file). Instead, your authorship is very well documented via the git history automatically.

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -1,6 +1,7 @@
 /* SPDX-License-Identifier: BSD-3-Clause */
 /*
  * Authors: Simon Kuenzer <simon.kuenzer@neclab.eu>
+ * 			Sairaj Kodilkar <skodilkar7@gmail.com>
```
Broken indentation. Use spaces in comments to indent. Align with other authors' names.

marcrittinghaus commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,87 @@
+#define uk_rwlock_init(rwl) __uk_rwlock_init(rwl, 0)
+
+#define uk_rwlock_init_config(rwl, config_flags) \
+	__uk_rwlock_init(rwl, config_flags)
```
What is the reason for the `__`-function name? Besides being discouraged by the coding convention, you can just define `uk_rwlock_init_config` as the actual main function, right?!

```C
void uk_rwlock_init_config(struct uk_rwlock *rwl, uint8_t config_flags);

#define uk_rwlock_init(rwl) uk_rwlock_init_config(rwl, 0)
```

marcrittinghaus commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,87 @@
+
+void uk_rwlock_downgrade(struct uk_rwlock *rwl);
+
+#define uk_rwlock_init(rwl) __uk_rwlock_init(rwl, 0)
```
Please add proper documentation for all functions in the following form. Make clear under what circumstances the function can be called. For example, when is it possible to up- or downgrade a lock? Can the function be called multiple times (i.e., recursive locking behavior)?

```C
/**
 * Does X and Y. Description of state, requirements, etc.
 *
 * @param rwl
 *   Reader-writer lock to operate on
 */
void uk_rwlock_XY(struct uk_rwlock *rwl);
```

marcrittinghaus commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,87 @@
+
+#include <uk/print.h>
+
+#define UK_RWLOCK_CONFIG_SPIN			0x01
```
Broken indentation

marcrittinghaus commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,87 @@
+	unsigned int config_flags;
+	struct uk_spinlock sl;
+	struct uk_waitq shared;
+	struct uk_waitq exclusive;
```
Please add short comments on the purpose of each field like so:

```C
/** Reader-Writer Lock */
struct __align(8) uk_rwlock {
	/** Number of active readers, -1 if writer active */
	int nactive;
	/** ... */
	int npending_reads;
        /** Configuration flags (see UK_RWLOCK_CONFIG_*) */
        unsigned int config_flags;
...
```

andraprs commented:
> Thank you for the updates from the latest revision and for developing and sharing a testing setup [1]. I checked out the patch series using the provided test app.
> 
> Reviewed-by: Andra Paraschiv [andra@unikraft.io](mailto:andra@unikraft.io)
> 
> [1] https://github.com/Sairajkodilkar/unikraft/commits/sched-sync

Thank you for including the updates for the license / author header(s) [2].

There are a couple of warnings in the checkpatch output mentioned above (e.g. with regard to the commit message), please check them out.

Thank you.

[2] https://github.com/unikraft/unikraft/compare/e3c28749859170980a0711e1ff83e9101c9a70f1..917439c13004431998bf2fb6fff46bd7dcd0b6f

marcrittinghaus commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,87 @@
+#include <uk/print.h>
+
+#define UK_RWLOCK_CONFIG_SPIN			0x01
+#define UK_RWLOCK_CONFIG_WRITE_RECURSE	0x02
```
Please add a short comment on the meaning of the config options like so:

```C
#define UK_RWLOCK_CONFIG_SPIN		0x01 /* ??? */
#define UK_RWLOCK_CONFIG_WRITE_RECURSE	0x02 /* Recursive locking for writers */
```

marcrittinghaus commented:
Please wait with further changes. I am currently doing a review, thanks 😃

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -90,61 +97,68 @@ extern __spinlock              _uk_mutex_metrics_lock;
 void uk_mutex_get_metrics(struct uk_mutex_metrics *dst);
 
+#define uk_mutex_init(m) __uk_mutex_init(m, 0)
+#define uk_mutex_init_config(m, config_flags) __uk_mutex_init(m, config_flags)
```
Same as for `rwlock`. Get rid of `__uk_mutex_init()`. Rename that function to `uk_mutex_init_config()`. Define `uk_mutex_init()` as shortcut

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -52,16 +54,21 @@
 extern "C" {
 #endif
 
+#define UK_MUTEX_CONFIG_WRITE_RECURSE 0x01
```
Does not make sense to have a `_WRITE_` in there, right? Just call it `UK_MUTEX_CONFIG_RECURSE`.

Also add a short comment:
```C
#define UK_MUTEX_CONFIG_RECURSE 0x01 /* ... */
```

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -52,16 +54,21 @@
 	struct uk_waitq wait;
 };
 
+#define __IS_CONFIG_WRITE_RECURSE(flag)	(flag & UK_MUTEX_CONFIG_WRITE_RECURSE)
```
With the new coding convention, we want to avoid `__` names for things like this. Actually, I find it totally fine to define a proper function for this:

```C
static inline int uk_mutex_is_recursive(struct uk_mutex *m)
{
	return (m->config_flags & UK_MUTEX_CONFIG_RECURSE);
}
```

Since the name is shorter and you don't have to specify the flags, you can also fit everything in one line for the `if` in line 117.

marcrittinghaus commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,87 @@
+	struct uk_waitq exclusive;
+};
+
+
```
Remove the excess new line.

marcrittinghaus commented on lib/uklock/include/uk/lock-common.h:
```diff
@@ -0,0 +1,63 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
Missing include guards

marcrittinghaus commented on lib/uklock/include/uk/lock-common.h:
```diff
@@ -0,0 +1,63 @@
+ */
+
+#include <uk/plat/lcpu.h>
+
```
Missing cplusplus block (below includes)

```C
#ifdef __cplusplus
extern "C" {
#endif
```

marcrittinghaus commented on lib/uklock/include/uk/spinlock.h:
```diff
@@ -37,6 +40,8 @@
 extern "C" {
 #endif
 
+#include <uk/lock-common.h>
```
Move outside the `cpluscplus` block

marcrittinghaus commented on lib/uklock/include/uk/semaphore.h:
```diff
@@ -45,6 +47,7 @@ extern "C" {
  * uses wait queues for threads
  */
 struct uk_semaphore {
+	uk_spinlock sl;
```
Add the necessary `#include <uk/spinlock.h>` to the list of includes.

marcrittinghaus commented on lib/uklock/include/uk/spinlock.h:
```diff
@@ -1,9 +1,12 @@
 /* SPDX-License-Identifier: BSD-3-Clause */
 /*
+ * Authors: Sairaj Kodilkar <skodilkar7@gmail.com>
+ *
```
Same as before.

marcrittinghaus commented on lib/uklock/include/uk/spinlock.h:
```diff
@@ -1,9 +1,12 @@
  * Copyright (c) 2021 Karlsruhe Institute of Technology (KIT).
  *               All rights reserved.
  *
+ *
```
Remove additional newline

marcrittinghaus commented on lib/uklock/mutex.c:
```diff
@@ -1,5 +1,38 @@
 #include <uk/mutex.h>
 
```
The includes come after the header. The first line must be the SPDX license identifier.

marcrittinghaus commented on lib/uklock/mutex.c:
```diff
@@ -1,5 +1,38 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Authors: Simon Kuenzer <simon.kuenzer@neclab.eu>
+ * 			Sairaj Kodilkar <skodilkar7@gmail.com>
```
Same as above

marcrittinghaus commented on lib/uksched/include/uk/wait.h:
```diff
@@ -1,5 +1,7 @@
 /* SPDX-License-Identifier: BSD-2-Clause */
 /*
+ * Authors: Sairaj Kodilkar <skodilkar7@gmail.com>
+ *
```
Same as before

marcrittinghaus commented on lib/uksched/include/uk/wait.h:
```diff
@@ -71,28 +74,33 @@ void uk_waitq_remove(struct uk_waitq *wq,
 	if (entry->waiting) {
-		UK_STAILQ_REMOVE(wq, entry, struct uk_waitq_entry, thread_list);
+		UK_STAILQ_REMOVE(&(wq->wait_list), entry,
+						struct uk_waitq_entry, thread_list);
```
```suggestion
		UK_STAILQ_REMOVE(&(wq->wait_list), entry,
				 struct uk_waitq_entry, thread_list);
```

marcrittinghaus commented on lib/uksched/include/uk/wait.h:
```diff
@@ -71,28 +74,33 @@ void uk_waitq_remove(struct uk_waitq *wq,
+/* TODO investigate uk_sched_thread_blocked
+ * is clear_runnable critical
+ * is uk_sched_thread_blocked critical
+ */
```
What needs to be done here? Can this be removed?

marcrittinghaus commented on lib/uksched/include/uk/wait.h:
```diff
@@ -126,15 +134,48 @@ do { \
+#define uk_waitq_wait_locked(wq, lock_fn, unlock_fn, lock) \
+	uk_waitq_wait_deadline(wq, lock_fn, unlock_fn, lock, 0, 0)
+
+
```
Remove excess newline

marcrittinghaus commented on lib/uksched/include/uk/wait_types.h:
```diff
@@ -1,5 +1,7 @@
 /* SPDX-License-Identifier: BSD-2-Clause */
 /*
+ * Authors: Sairaj Kodilkar <skodilkar7@gmail.com>
+ *
```
Same here

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -90,61 +97,68 @@ extern __spinlock              _uk_mutex_metrics_lock;
-			m->lock_count == 0 || m->owner == current);
-		irqf = ukplat_lcpu_save_irqf();
-		if (m->lock_count == 0 || m->owner == current)
+
```
Remove excess newline

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -90,61 +97,68 @@ extern __spinlock              _uk_mutex_metrics_lock;
 
 	current = uk_thread_current();
 
+	/* If the owner is a current thread, then we increment the lock count. */
```
Exceeds 80 columns.

```suggestion
	/* If the owner is the current thread we increment the lock count */
```

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -90,61 +97,68 @@ extern __spinlock              _uk_mutex_metrics_lock;
+
+		/* If there is no owner then we can acquire the lock and increment the
+		 * lock count.
+		 */
```
Exceeds 80 columns
```suggestion
		/* If there is no owner we can acquire the lock
		 * and increment the lock count
		 */
```

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -90,61 +97,68 @@ extern __spinlock              _uk_mutex_metrics_lock;
+		/* If there is no owner then we can acquire the lock and increment the
+		 * lock count.
+		 */
+		if (ukarch_compare_exchange_sync(&m->owner, 0, current) == current) {
```
Exceeds 80 columns. Use `NULL` for pointers.
Rename `current` -> `cur` in this function and `uk_mutex_trylock`.

```suggestion
		if (ukarch_compare_exchange_sync(&m->owner, NULL, cur) == cur) {
```

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -90,61 +97,68 @@ extern __spinlock              _uk_mutex_metrics_lock;
+		return;
+	} else {
+		UK_ASSERT(m->owner != current);
+	}
```
Make more obvious that there is no way around this assert if you continue to execute the function.
```suggestion
	}
	
	UK_ASSERT(m->owner != current);
```

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -90,61 +97,68 @@ extern __spinlock              _uk_mutex_metrics_lock;
+	/* If the owner is a current thread, then we increment the lock count. */
+	if (__IS_CONFIG_WRITE_RECURSE(m->config_flags)
+			&& unlikely(m->owner == current)) {
+		ukarch_inc(&(m->lock_count));
```
Why do we need an atomic increment here? Since we are the owner of the lock, no other thread will increment or decrement this count. The only situation could be an interrupt. But we should not try to acquire a blocking synchronization primitive in an interrupt handler anyways.

Sairajkodilkar replied:
Yes you are right, we dont need an atomic increment

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -90,61 +97,68 @@ extern __spinlock              _uk_mutex_metrics_lock;
+		 * lock count.
+		 */
+		if (ukarch_compare_exchange_sync(&m->owner, 0, current) == current) {
+			ukarch_inc(&(m->lock_count));
```
Same question concerning the atomic needed or not. If not, I would even suggest the following to save the read.
```suggestion
			UK_ASSERT(m->lock_count == 0);
			m->lock_count = 1;
```

marcrittinghaus commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,87 @@
+struct __align(8) uk_rwlock {
+	int nActive;
+	int nPendingReads;
+	int nPendingWrites;
```
We use snake_case.
```suggestion
	int nactive;
	int npending_reads;
	int npending_writes;
```

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -52,16 +54,21 @@
  */
 struct uk_mutex {
 	int lock_count;
+	uint8_t config_flags;
```
Due to padding you do not gain any savings by using `uint8_t`. In fact, since you have to use longer `movzx` instructions to zero-extend the value during load, it can happen that the OPCODE becomes even longer. Although that depends on the code. 

Nevertheless, since you are using an `unsigned int` in `uk_rwlock`, I would use the same here for consistency.

```suggestion
	unsigned int config_flags;
```

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -90,61 +97,68 @@ extern __spinlock              _uk_mutex_metrics_lock;
+	{ 0, 0, NULL, __WAIT_QUEUE_INITIALIZER((name).wait) }
 
-void uk_mutex_init(struct uk_mutex *m);
+void __uk_mutex_init(struct uk_mutex *m, uint8_t config_flags);
```
Reminder: use `unsigned int config_flags` here, too, if you change the type in the `struct`.

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -90,61 +97,68 @@ extern __spinlock              _uk_mutex_metrics_lock;
-	m->owner = current;
 
-#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
 	_uk_mutex_metrics.active_locked   += (m->lock_count == 1);
 	_uk_mutex_metrics.active_unlocked -= (m->lock_count == 1);
```
`m->lock_count == 1` is always true.

Btw: It is only guaranteed that true is a non-zero value. But it does not need to be `1`. So this was already broken before.

Sairajkodilkar replied:
Hi Marc, I am thinking to create a new PR to fix this issue, as the bug is not introduced by my changes

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -90,61 +97,68 @@ extern __spinlock              _uk_mutex_metrics_lock;
 
-	irqf = ukplat_lcpu_save_irqf();
-	if (m->lock_count == 0 || m->owner == current) {
+	if (__IS_CONFIG_WRITE_RECURSE(m->config_flags) && m->owner == current)
```
This looks like a lot of unnecessary operations. There also seem to be some bugs in the existing code. The metrics miss locking and the `unlocked` is incremented instead of decremented. 

What about:
```C
static inline int uk_mutex_trylock(struct uk_mutex *m)
{
	struct uk_thread *cur;

	UK_ASSERT(m);

	cur = uk_thread_current();

	if (unlikely(uk_mutex_is_recursive(m) && m->owner == cur)) {
		m->lock_count++;

#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
		ukarch_spin_lock(&_uk_mutex_metrics_lock);
		_uk_mutex_metrics.total_ok_trylocks++;
		ukarch_spin_unlock(&_uk_mutex_metrics_lock);
#endif /* CONFIG_LIBUKLOCK_MUTEX_METRICS */

		return 1;
	}

	UK_ASSERT(m->owner != cur);

	if (ukarch_compare_exchange_sync(&m->owner, NULL, cur) == cur) {
		UK_ASSERT(m->lock_count == 0);
		m->lock_count = 1;

#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
		ukarch_spin_lock(&_uk_mutex_metrics_lock);
		_uk_mutex_metrics.active_locked++;
		_uk_mutex_metrics.active_unlocked--;
		_uk_mutex_metrics.total_ok_trylocks++;
		ukarch_spin_unlock(&_uk_mutex_metrics_lock);
#endif /* CONFIG_LIBUKLOCK_MUTEX_METRICS */

		return 1;
	}

#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
	ukarch_spin_lock(&_uk_mutex_metrics_lock);
	_uk_mutex_metrics.total_failed_trylocks++;
	ukarch_spin_unlock(&_uk_mutex_metrics_lock);
#endif /* CONFIG_LIBUKLOCK_MUTEX_METRICS */

	return 0;
}
```

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -90,61 +97,68 @@ extern __spinlock              _uk_mutex_metrics_lock;
 
+	/* If the owner is a current thread, then we increment the lock count. */
+	if (__IS_CONFIG_WRITE_RECURSE(m->config_flags)
+			&& unlikely(m->owner == current)) {
```
I think it is safe to assume that recursive locks are less used. So if you want to use the `unlikely` attribute, I would put it around the whole condition.
```suggestion
	if (unlikely(uk_mutex_is_recursive(m) && m->owner == cur)) {
```

marcrittinghaus commented on lib/uklock/include/uk/rwlock.h:
```diff
@@ -0,0 +1,87 @@
+#include <uk/lock-common.h>
+#include <uk/thread.h>
+
+#include <uk/print.h>
```
Only include headers that are really needed for definitions in this file. All other includes go into the `*.c` file.

marcrittinghaus replied:
Missing cplusplus block after the includes
```C
#ifdef __cplusplus
extern "C" {
#endif
```

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -153,29 +167,30 @@ static inline int uk_mutex_trylock(struct uk_mutex *m)
 	UK_ASSERT(m->lock_count > 0);
-	if (--m->lock_count == 0) {
+
+	if (ukarch_sub_fetch(&(m->lock_count), 1) == 0) {
```
Again, I don't think, we really need atomic instructions here. You just need to make sure you use a memory barrier.
```suggestion
	if (--m->lock_count == 0) {
		/* Make sure lock_count is visible before resetting the
		 * owner. The lock can be acquired afterwards.
		 */
		wmb();
		...
```

marcrittinghaus commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -153,29 +167,30 @@ static inline int uk_mutex_trylock(struct uk_mutex *m)
-
 #ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
 	ukarch_spin_lock(&_uk_mutex_metrics_lock);
 #endif /* CONFIG_LIBUKLOCK_MUTEX_METRICS */
```
Although not part of your changes, it is a pretty bad idea to keep the stats spinlock while calling `uk_waitq_wake_up()`. The scheduler is invoked in there and depending on the implementation (e.g., priority based) it can immediately switch threads. If the woken thread then tries to acquire/release **any** lock, we run into a deadlock! ⚠️⚠️⚠️

Please update the code to acquire the spinlock after the fact like so:
```C
if (ukarch_sub_fetch(&(m->lock_count), 1) == 0) {
	m->owner = NULL;
	uk_waitq_wake_up(&m->wait);

#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
	ukarch_spin_lock(&_uk_mutex_metrics_lock);
	_uk_mutex_metrics.active_locked--;
	_uk_mutex_metrics.active_unlocked++;
	_uk_mutex_metrics.total_unlocks++;
	ukarch_spin_unlock(&_uk_mutex_metrics_lock);
#endif /* CONFIG_LIBUKLOCK_MUTEX_METRICS */
} else {
#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
	ukarch_spin_lock(&_uk_mutex_metrics_lock);
	_uk_mutex_metrics.total_unlocks++;
	ukarch_spin_unlock(&_uk_mutex_metrics_lock);
#endif /* CONFIG_LIBUKLOCK_MUTEX_METRICS */
}
```

marcrittinghaus commented on lib/uklock/rwlock.c:
```diff
@@ -0,0 +1,153 @@
+	uk_waitq_init(&rwl->exclusive);
+}
+
+
```
Remove excess newline

marcrittinghaus commented on lib/uklock/rwlock.c:
```diff
@@ -0,0 +1,153 @@
+	uk_spin_unlock(&rwl->sl);
+}
+
+
```
Remove excess newline

marcrittinghaus commented on lib/uklock/rwlock.c:
```diff
@@ -0,0 +1,153 @@
+	uk_spin_lock(&rwl->sl);
+	rwl->nPendingReads++;
+
+
```
Remove excess newline

marcrittinghaus commented on lib/uklock/rwlock.c:
```diff
@@ -0,0 +1,153 @@
+#include <uk/config.h>
+
+#define __IS_CONFIG_SPIN(flag)			(flag & UK_RWLOCK_CONFIG_SPIN)
+#define __IS_CONFIG_WRITE_RECURSE(flag)	(flag & UK_RWLOCK_CONFIG_WRITE_RECURSE)
```
Avoid `__`-names. Just use `RWLOCK_IS_CONFIG_SPIN` etc. without the `uk` prefix to signal that it is private.

+ Indentation broken

marcrittinghaus commented on lib/uklock/rwlock.c:
```diff
@@ -0,0 +1,153 @@
+				&rwl->shared,
+				uk_spin_lock,
+				uk_spin_unlock,
+				&rwl->sl);
```
Please break the arguments like so. We had this topic in the last maintainers meeting for the coding convention.
```suggestion
		uk_waitq_wait_locked(&rwl->shared,
				     uk_spin_lock,
				     uk_spin_unlock,
				     &rwl->sl);
```

marcrittinghaus commented on lib/uklock/rwlock.c:
```diff
@@ -0,0 +1,153 @@
+void uk_rwlock_runlock(struct uk_rwlock *rwl)
+{
+	uk_spin_lock(&rwl->sl);
+	rwl->nActive--;
```
I suggest to add `UK_ASSERT`s in the code at various points to make assumptions you have explicit, like here `UK_ASSERT(rwl->nActive > 0);`, or for `uk_rwlock_wunlock()` you should add a `UK_ASSERT(rwl->nActive == -1);`.

marcrittinghaus commented on lib/uklock/rwlock.c:
```diff
@@ -0,0 +1,153 @@
+		uk_waitq_wake_up_one(&rwl->exclusive);
+	} else {
+		uk_spin_unlock(&rwl->sl);
+	}
```
Generally, it is good to structure similar functions in a similar way. In the other functions you save the result from the condition to a variable and perform the check outside the lock. I do not particularly prefer one over the other but it simplifies comprehension and maintenance of the code.

marcrittinghaus commented on lib/uklock/rwlock.c:
```diff
@@ -0,0 +1,153 @@
+				&rwl->sl);
+
+	/* Wait for existing write to unlock the lock
+	 * The separate waits for the both ensures that there is no starvation
```
I do understand your idea here and I find it to be a clever solution. I think it would be good to have a separate comment block at the beginning of the source file after the header to describe how the rwlock works. In particular, when writers are prioritized over readers and the other way around and how starvation is prevented.

```C
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * Reader-Writer Lock
 * <DESCRIPTION HERE>
 */

#include <uk/rwlock.h>
```

marcrittinghaus commented on lib/uklock/rwlock.c:
```diff
@@ -0,0 +1,153 @@
+void uk_rwlock_upgrade(struct uk_rwlock *rwl)
+{
+	uk_spin_lock(&rwl->sl);
+	if (rwl->nActive == 1) {
```
Also here, I would add a `UK_ASSERT` to make sure you are not trying to upgrade a writer lock. Same for the downgrade etc.

marcrittinghaus commented on lib/uksched/include/uk/wait_types.h:
```diff
@@ -38,10 +41,16 @@ struct uk_waitq_entry {
+};
 
-#define __WAIT_QUEUE_INITIALIZER(name) UK_STAILQ_HEAD_INITIALIZER(name)
+#define __WAIT_QUEUE_INITIALIZER(name)	\
+{ \
+	UK_SPINLOCK_INITIALIZER(), \
+	UK_STAILQ_HEAD_INITIALIZER(name.wait_list) \
+}
```
Please indent the definition. And also uniformly use a single whitespace here before the `\` (first line uses a tab).

Sairajkodilkar replied:
Hi @marcrittinghaus, Could you explain what you mean by indenting the definition?

marcrittinghaus commented on lib/uksched/include/uk/wait.h:
```diff
@@ -126,15 +134,48 @@ do { \
 	timedout; \
 })
 
+#define uk_waitq_wait_deadline(wq, lock_fn, unlock_fn, lock, deadline, deadline_condition) \
```
Exceeds 80 columns

marcrittinghaus commented on lib/uksched/include/uk/wait.h:
```diff
@@ -126,15 +134,48 @@ do { \
+	uk_sched_yield(); \
+	if (lock) \
+		lock_fn(lock); \
+	if (deadline_condition) { \
+		timedout = 1; \
+	} \
```
Remove `{` `}`

marcrittinghaus reviewed (changes requested):
Hi @Sairajkodilkar,

Many thanks for putting this together and also for doing multiple iterations already 😃 I went through the code and unfortunately still have a bunch of comments. Generally, all of them should be easy and quick to fix.

Are you able to provide an update till next week? We really would like to upstream this now.

andraprs commented:
Hi Sairaj, Marc,

As a heads-up, checked out the latest version of the "synchronization" branch [1], per a discussion with Marc today to try it out. I added the last 3 commits [2][3][4] from the "sched-sync" branch [5] on top. May this setup needs updates, just let me know.

Sairaj, can you please rebase the "synchronization" branch on top of the latest "staging" branch version.

Between different runs of the sync-test app, there are different outputs, with a hang / assert as outcome. Here are a couple of examples:

```
dbg:  [apphelloworld] <main.c @   45> count is 10000000000
dbg:  [apphelloworld] <main.c @   45> count is 10000000000
```
```
dbg:  [apphelloworld] <main.c @   45> count is 20000000000
```
```
dbg:  [apphelloworld] <main.c @   45> count is 0
```
```
dbg:  [apphelloworld] <main.c @   45> count is 10000000000
CRIT: [libuklock] <rwlock.c @   72> Assertion failure: rwl->nactive > 0
```

Thank you.

[1] https://github.com/Sairajkodilkar/unikraft/commits/synchronization
[2] https://github.com/Sairajkodilkar/unikraft/commit/4c752a30e536ecaf63a8462058d029799b764d85
[3] https://github.com/Sairajkodilkar/unikraft/commit/8aace8e6222067efa479e26b8bf1576530c922a3
[4] https://github.com/Sairajkodilkar/unikraft/commit/0c57d5e1637b81bf83d861b7ec9a10b3f719e413
[5] https://github.com/Sairajkodilkar/unikraft/commits/sched-sync

marcrittinghaus commented:
Thanks @Sairajkodilkar for updating the PR 🙏🏻 😃 Please look if you can identify what change causes the issues. Regarding my comments, as I said, I am not sure if the we need the atomic operations at the mentioned places. I would say no, but I didn't test it. That is why I was asking to understand your rational behind using atomic instructions there. 

I also found a couple alignment problems (tabs vs. spaces, tabs = 8 spaces) still. Please make sure that these are fixed as well as the checkpatch issues are resolved.

I will revisit the PR and do another pass afterwards.

Thanks @andraprs for testing! 😃

marcrittinghaus commented:
@Sairajkodilkar I have done quite a lot of changes in the end. The most important are listed here:

- Removed changes that were undone by later commits
- Squashed some of the smaller commits (e.g., add rwlock to Makefile)
- Fixed the synchronization bug in the `rwlock` introduced by the latest changes
- Fixed a lot of style issues (indentation, function documentation style, etc.)
- Removed the common `irqf` interface. From my perspective, I does not make sense to have something an IRQ handler might access protected by a blocking synchronization primitive. It is different for incrementing a semaphore.
- Changed the `uk_spin_lock_irqf` functions to match the corresponding (already existing) `ukplat_spin_lock_irqsave`/`irqrestore` functions
- Switched from `uk_spinlock` to `__spinlock` for the wait queues and accordingly now use `ukplat_spin_lock_*`. This is because `uklock` depends on `uksched`, but it should not be the other way round.
- Added comments in `rwlock.c` to describe how the `rwlock` works.

**For reviewers:** ⚠️
The synchronization primitives still need `thread.c` and the scheduler to be properly synchronized. Otherwise, race conditions will happen when a lock wakes up a waiting thread of a foreign core and thereby accesses the `uk_thread` structure and scheduler queues in parallel to the foreign core.

andraprs commented:
I just checked the test app from the "sched-sync" branch [1] with the latest "staging" branch and the commits from this PR on top. It worked fine. Thank you.

```
[    5.608145] dbg:  [apphelloworld] <main.c @   45> count is 20000000000
[    5.657920] dbg:  [apphelloworld] <main.c @   45> count is 20000000000
[    5.701035] Warn: [apphelloworld] <main.c @  105> Test Passed
```

[1] https://github.com/Sairajkodilkar/unikraft/commits/sched-sync

marcrittinghaus commented:
@andraprs Thanks a lot for testing 😃 

@Sairajkodilkar Are you ok with the changes? We could merge then.

Sairajkodilkar commented:
Hi @marcrittinghaus, I tested the PR with the changes that you suggested. Thanks for helping out with it. 
As discussed we need to synchronize scheduler to be able to use this PR

andraprs reviewed (approved):
Reviewed-by: Andra Paraschiv <andra@unikraft.org>

Thank you.

marcrittinghaus commented:
For anyone who wants to test. This patch synchronizes the thread structure (to some degree). It is still not complete, but reduces the chances of races during testing.

[thread.c.patch](https://github.com/unikraft/unikraft/files/11469542/thread.c.patch)

RaduNichita commented:
I tested the app and it seemed to work fine for me.

Reviewed-by: Radu Nichita <radunichita99@gmail.com>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
